samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -705,43 +714,42 @@ public void run() {
     this.sideInputSystemConsumers.start();
 
     // create a thread for sideInput reads
-    Thread readSideInputs = new Thread(() -> {
-        while (!shutDownSideInputRead) {
-          IncomingMessageEnvelope envelope = sideInputSystemConsumers.choose(true);
-          if (envelope != null) {
-
-            if (!envelope.isEndOfStream())
-              sideInputStorageManagers.get(envelope.getSystemStreamPartition()).process(envelope);
+    Thread readSideInputsThread = new Thread(() -> {
+        try {
+          while (!shutDownSideInputRead) {
+            IncomingMessageEnvelope envelope = sideInputSystemConsumers.choose(true);
+            if (envelope != null) {
 
-            checkSideInputCaughtUp(envelope.getSystemStreamPartition(), envelope.getOffset(),
-                SystemStreamMetadata.OffsetType.NEWEST, envelope.isEndOfStream());
+              if (!envelope.isEndOfStream())
+                sideInputStorageManagers.get(envelope.getSystemStreamPartition()).process(envelope);
 
-          } else {
-            LOG.trace("No incoming message was available");
+              checkSideInputCaughtUp(envelope.getSystemStreamPartition(), envelope.getOffset(),
```

Comment:
not necessarily in this PR but should re-factor checkSideInputCaughtup since it's extremely complex.

a few thoughts on this:
- should return a boolean from the API
- too many params that are un-related (eg: a few fields from envelope and an offset-type)
- signature should be `boolean isCaughtUp(envelope)`
- should surface the count-down latch mutation in a more obvious manner instead of happening 2 levels downstream

Reply:
Sure, this method is mostly akin to TaskInstance.checkCaughtUp. 
I will try to simplify this as a part of SAMZA-2257